Add possibility to store results of a simulation

camillegontier opened the pull request:
Modifications to retrieve the results of the simulation in a Results structure. At each time step, the following values are retrieved:
- MAP (to be able to compute the inference errors)
- Inter stimulation times (ISI) 
- Computation times for propagate!(sim)
- entropies for all 5 parameters

camillegontier committed:
Retrieve the entropy at each time step

Add a save_results! function to retrieve the value of the entropy at each time step

camillegontier committed:
Debug

Debug and typos correction

camillegontier committed:
Debug

camillegontier committed:
Define a Results structure

Define a Results structure to retrieve results at each time step

camillegontier committed:
Typo correction

camillegontier committed:
Add entropy

Add entropy function to retrieve the entropy of the posterior at each time step

camillegontier committed:
Add JLD package

Add JLD package for saving the results at each time step

camillegontier committed:
Add _tauentropy functions

Add _tauentropy functions for optimizing only the entropy of the posterior distribution of tau

camillegontier committed:
Add Myopic_tau and MyopicFast_tau policies

camillegontier committed:
Add Myopic_tau and MyopicFast_tau

camillegontier committed:
Debug

camillegontier committed:
Debug

camillegontier committed:
Debug

camillegontier committed:
Add condition isdefined(ENV["SLURM_ARRAY_TASK_ID"])

Add condition isdefined(ENV["SLURM_ARRAY_TASK_ID"]) : only record the results if a job array on Ubelix is launched

camillegontier committed:
Save ISI

Record the ISI at the end of the simulation

camillegontier committed:
Update simulate.jl

Retrieve the MAP values, entropies and computation times at each step of the simulation

camillegontier committed:
Update simulate.jl

Debugging

camillegontier committed:
Update simulate.jl

Debug

camillegontier committed:
Retrieve computation times

simsurace commented on src/simulate.jl:
```diff
@@ -69,18 +91,80 @@ function propagate!(sim::NestedFilterSimulation, dt)
+    results.map.map_q[i] = map.q
+    results.map.map_σ[i] = map.σ
+    results.map.map_τ[i] = map.τ
+
```
Add `return` statement, otherwise the function will return the last result, i.e. `map.τ` in this case. I would suggest `return results`.

camillegontier replied:
Added return results statement

camillegontier committed:
Add return results statement

simsurace commented on src/simulate.jl:
```diff
@@ -69,18 +91,80 @@ function propagate!(sim::NestedFilterSimulation, dt)
         end
+        save_results!(results, sim, i, time)
+        if i == T
+            save(string(Base.parse(Int, ENV["SLURM_ARRAY_TASK_ID"]),".jld"), "entropies", results.entropies, "ISI", sim.times, "MAP", results.map)
```
This is not going to generalize to different machines/clusters and will throw errors if I run it on my machine.
It's better to define a general interface and keep specific things for the user scripts (which are not part of the package). I would suggest something like this
```julia
struct Recording{T1, T2, T3}
    f1::T1
    f2::T2
    data::T3
end

function Recording(f1, f2, sim::NestedFilterSimulation)
    res = f1(sim)
    data = [res]
    return Recording(f1, f2, data)
end

function update!(rec::Recording, sim)
    push!(rec.data, rec.f1(sim))
    return rec
end

function save(rec::Recording)
    rec.f2(rec.data)
end
```
You can then modify `run!` with an additional optional argument, i.e.
```julia
const NoRecording = Recording(x -> nothing, x -> nothing, Nothing[])

function run!(sim::NestedFilterSimulation; T::Int, plot_each_timestep::Bool = false, recording::Recording = NoRecording)
# modify code appropriately, using `update!` and `save` functions defined above
```
The user can then instantiate `Recording` and pass it to `run!`, e.g.
```julia
sim = NestedFilterSimulation(...)
f1(sim) = ... # code that computes entropies, maps, etc., whatever you need to record
f2(a) = ... # code that writes array `a` to a specific file, or if you don't want to do anything you type f2(a) = nothing, since you can then work with rec after running the simulation

rec = Recording(f1, f2, sim)

run!(sim, ...., recording = rec)
```

simsurace replied:
Better to put this into a separate file record.jl, to keep things easier to understand.

camillegontier replied:
I modified simulate.jl and created record.jl accordingly

simsurace commented on src/myopic.jl:
```diff
@@ -186,3 +188,58 @@ function _entropy(model::BinomialGridModel, obs::BinomialObservation, ::MyopicFa
     return argmin(entropies)
 end
+
+function _tauentropy(model::BinomialGridModel, obs::BinomialObservation, ::Myopic)
```
Could you add some unit tests for _tauentropy, similar to [here](https://github.com/Theoretical-Neuroscience-Group/BinomialSynapses.jl/blob/OED_HL_test_CG/test/myopic.jl#L42-L138)

camillegontier replied:
Done, if I understood correctly, results of the test should be the same for _tauentropy and _entropy, is that correct ?

simsurace commented on src/simulate.jl:
```diff
@@ -8,6 +8,28 @@ struct NestedFilterSimulation{T1, T2, T3, T4, T5, T6, T7}
     epsps::T7
 end
 
+struct Map{T1, T2, T3, T4, T5}
```
You won't need `Map`, `Entropies`, and `Results` anymore after implementing `Recording` (see other comment).

camillegontier replied:
Structures removed (as well as function save_results)

camillegontier committed:
Create record.jl

Linked to comment https://github.com/Theoretical-Neuroscience-Group/BinomialSynapses.jl/pull/37#discussion_r742942189

camillegontier committed:
Add include("record.jl")

camillegontier committed:
Add const NoRecording

simsurace commented:
Awesome, will try to run tests tomorrow on my local machine with GPU.

camillegontier committed:
Add update! and save to the run! function

camillegontier committed:
Add arguments i and time to function update!

camillegontier committed:
Remove structures Map, Entropies, Results, and function save_results!

simsurace commented on src/record.jl:
```diff
@@ -10,8 +10,8 @@ function Recording(f1, f2, sim::NestedFilterSimulation)
 
-function update!(rec::Recording, sim)
-    push!(rec.data, rec.f1(sim))
+function update!(rec::Recording, sim, i, time)
```
What is `i` needed for? I can see that `time` is needed because that is not stored in `sim`.

camillegontier replied:
It is meant to modify the vector of stored values, something like 
my_values = zertos(T)
rec.my_values[i] = sim.value[i]
Shall we use something more efficient, like a "append" function ?

simsurace replied:
`update!` uses `push!`, so this is not needed

camillegontier committed:
Add test for _tauentropy

simsurace committed:
Remove JLD from dependencies

simsurace committed:
Define Recording before simulate.jl

simsurace committed:
Fix Recording

All tests passing

simsurace committed:
Change NoRecording from type to const

simsurace committed:
Add unit tests for record.jl, fix initialization of Recording

simsurace committed:
Remove redundant methods for update_data!

simsurace committed:
Remove export of `save` and `update!`

simsurace committed:
Include Manifest.toml

simsurace committed:
Switch to older version of BinomialGPU

simsurace committed:
Merge pull request #38 from Theoretical-Neuroscience-Group/OED_HL_test_simone

Test and fix everything up
